Plugins/Youbora/YouboraAdnalyzerManager.swift

```diff
@@ -0,0 +1,201 @@
+
+extension YouboraAdnalyzerManager {
+    
+    private var adEventsToRegister: [AdEvent.Type] {
+        return [
+            AdEvent.adLoaded,
+            AdEvent.adStarted,
+            AdEvent.adComplete,
+            AdEvent.adResumed,
+            AdEvent.adPaused,
+            AdEvent.adDidProgressToTime,
+            AdEvent.adSkipped,
+            AdEvent.adStartedBuffering,
+            AdEvent.adPlaybackReady,
+            AdEvent.adsRequested
+        ]
+    }
+    
+    fileprivate func registerAdEvents(onMessageBus messageBus: MessageBus) {
+        PKLog.debug("register ad events")
+        
+        self.adEventsToRegister.forEach { event in
+            PKLog.debug("\(String(describing: type(of: self))) will register event: \(event.self)")
+            
+            switch event {
+            case let e where e.self == AdEvent.adLoaded:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    // update ad info with the new loaded event
+                    self?.adInfo = event.adInfo
+                    // if ad is preroll make sure to call /start event before /adStart
+                    if let positionType = event.adInfo?.positionType, positionType == .preRoll {
+                        self?.plugin.playHandler()
+                    }
+                    self?.playAdHandler()
+                }
+            case let e where e.self == AdEvent.adStarted:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    self?.joinAdHandler()
+                }
+            case let e where e.self == AdEvent.adComplete:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    self?.endedAdHandler()
+                    self?.adInfo = nil
+                }
+            case let e where e.self == AdEvent.adResumed:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    self?.resumeAdHandler()
+                }
+            case let e where e.self == AdEvent.adPaused:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    self?.pauseAdHandler()
+                }
+            case let e where e.self == AdEvent.adDidProgressToTime:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    // update ad playhead with new data
+                    self?.adPlayhead = event.adMediaTime?.doubleValue
+                }
+            case let e where e.self == AdEvent.adSkipped:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    self?.skipAdHandler()
+                }
+            case let e where e.self == AdEvent.adStartedBuffering:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    self?.bufferingAdHandler()
```

Comment:
The delegate should only work for in the middle buffering after the ad started playing, thats what google is describing in their scenario. I will test to make sure but if it doesn't work like this it is something that should be dealt with IMA SDK behaviors. I took the semantic from IMA, ended sounds like no more buffering is going to happen which is incorrect we want to tell the developer that enough buffering was made and playback should keep playing without delays.